### PR TITLE
Mu8/javier/issue13 create characters detail activity

### PR DESCRIPTION
**Pull request by jamacado**
**Create Characters Detail Activity #13**

Anotación 2020-02-21 223453

- create characters detail activity
- create adapter for comics by character

**Testing Done**

- Nexus 6P

**How to test:**

- Run the application and click en character, select one character.

**Signed-off-by:**

- Javier M. Cardona javier.mcardona@gmail.com

### CODE REVIEW

File: data/src/main/java/com/architect/coders/mu8/data/charactersDetail/CharacterDetailRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+class CharacterDetailRepositoryImpl(application: DataApp) : CharacterDetailRepository {
+
+    private val database = application.database
+
+    override suspend fun getComicsForCharacter(id: Long, idComics: List<String>): List<Comic> = withContext(Dispatchers.IO) {
+        with(database.getComicsDao()) {
+            var listComicsforCharacter = getListComicCharacter(idComics).map { it.toDomainModel() }
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/charactersDetail/CharacterDetailRepositoryImpl.kt]: Rename to `characterComics`

File: data/src/main/java/com/architect/coders/mu8/data/comics/ComicsDAO.kt [hunk 2]
```diff
@@ -17,6 +17,9 @@ interface ComicsDAO {
     @Insert(onConflict = IGNORE)
     fun insertComics(comics: List<ComicsEntity>)
 
-    @Query("SELECT characters_id FROM comics_table WHERE id = :comicId" )
-    fun getComicCharacters(comicId : Long) : List<String>
+    @Query("SELECT characters_id FROM comics_table WHERE id = :comicId")
+    fun getComicCharacters(comicId: Long): List<String>
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/comics/ComicsDAO.kt]: What is this query used for? Can you explain?

File: data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt [hunk 3]
```diff
@@ -45,4 +47,22 @@ interface MarvelServiceAPI {
         @Query(OFFSET) offset: Int,
         @Query(LIMIT) limit: Int
     ): Response<BaseResponse<EventsResponse>>
+
+    @GET("/v1/public/characters/{characterId}")
+    suspend fun findById(
```
Comment (recaldev) [inline data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt]: find what by id? Please give a better name for this function.

File: presentation/build.gradle [hunk 4]
```diff
@@ -48,6 +48,8 @@ dependencies {
     implementation SupportDependencies.firebaseAnalytics
 
     implementation LibsDependencies.glide
+    implementation 'androidx.appcompat:appcompat:1.1.0'
+    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
```
Comment (recaldev) [inline presentation/build.gradle]: Don't hardcode this dependencies

File: presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailActivity.kt [hunk 5]
```diff
@@ -1,14 +1,55 @@
...
+    private lateinit var viewModel: CharactersDetailViewModel
+    private lateinit var binding: ActivityCharactersDetailBinding
+    private lateinit var adapter: CharacterDetailAdapter
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        setContentView(R.layout.activity_characters_detail)
+        binding = DataBindingUtil.setContentView(this, R.layout.activity_characters_detail)
+
+        initVM(intent.getLongExtra(CHARACTER, -1))
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailActivity.kt]: Please be specific with the method names, what is VM?

Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailActivity.kt]: Please create a constant with a meaningful name for this value `-1`

File: presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailViewModel.kt [hunk 6]
```diff
@@ -1,14 +1,53 @@
...
+    }
+
+    private fun getDetailCharacter() {
+        launch {
+            _loading.value = true
+            try {
+                _character.value = charactersUseCase.findCharacter(characterId)
+                getComics()
+            } catch (e: IOException) {
+                // _detailCharacter.value = In
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailViewModel.kt]: Why id this commented? Why do you have a try catch?

File: presentation/src/main/res/layout/activity_characters_detail.xml [hunk 7]
```diff
@@ -1,12 +1,85 @@
...
-
-    <include
-        android:id="@+id/toolbar"
-        layout="@layout/toolbar" />
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+    xmlns:tools="http://schemas.android.com/tools">
+
+    <data>
+
```
Comment (recaldev) [inline presentation/src/main/res/layout/activity_characters_detail.xml]: Remove extra spaces please

File: presentation/src/main/res/layout/view_comic.xml [hunk 8]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools">
+
+    <data>
+        <variable
+            name="comic"
+            type="com.architect.codes.mu8.comics.Comic" />
+
```
Comment (recaldev) [inline presentation/src/main/res/layout/view_comic.xml]: Same here about the spaces

File: presentation/src/main/res/layout/activity_characters_detail.xml [hunk 9]
```diff
@@ -1,12 +1,85 @@
...
+            android:id="@+id/recycler_comic"
+            android:layout_width="0dp"
+            android:layout_height="@dimen/dp_200"
+            android:layout_marginStart="@dimen/margin_14"
+            android:layout_marginTop="@dimen/dp_20"
+            android:layout_marginEnd="@dimen/margin_14"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/tv_detail" />
+
```
Comment (recaldev) [inline presentation/src/main/res/layout/activity_characters_detail.xml]: Remove space

**Review by recaldev — CHANGES_REQUESTED** [review]
Please solve the comments

File: data/src/main/java/com/architect/coders/mu8/data/charactersDetail/CharacterDetailRepositoryImpl.kt (hunk 1 above)
Comment (jamacado) [reply data/src/main/java/com/architect/coders/mu8/data/charactersDetail/CharacterDetailRepositoryImpl.kt]: ok

File: data/src/main/java/com/architect/coders/mu8/data/comics/ComicsDAO.kt (hunk 2 above)
Comment (jamacado) [reply data/src/main/java/com/architect/coders/mu8/data/comics/ComicsDAO.kt]: was already created in issue-32, look develop

File: data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt (hunk 3 above)
Comment (jamacado) [reply data/src/main/java/com/architect/coders/mu8/data/service/MarvelServiceAPI.kt]: findCharacterById

File: presentation/build.gradle (hunk 4 above)
Comment (jamacado) [reply presentation/build.gradle]: delete lines

File: presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailActivity.kt (hunk 5 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailActivity.kt]: vm = viewmodel, change by initViewModel

Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/characters/detail/CharactersDetailActivity.kt]: ok

File: presentation/src/main/res/layout/activity_characters_detail.xml (hunk 7 above)
Comment (jamacado) [reply presentation/src/main/res/layout/activity_characters_detail.xml]: delete extra space

File: presentation/src/main/res/layout/view_comic.xml (hunk 8 above)
Comment (jamacado) [reply presentation/src/main/res/layout/view_comic.xml]: ctrl+alt+l make the space

File: presentation/src/main/res/layout/activity_characters_detail.xml (hunk 9 above)
Comment (jamacado) [reply presentation/src/main/res/layout/activity_characters_detail.xml]: remove space

File: presentation/src/main/res/layout/view_comic.xml (hunk 8 above)
Comment (recaldev) [reply presentation/src/main/res/layout/view_comic.xml]: Please remove it

**Review by recaldev — APPROVED** [review]

**Pull request merged**